src/shell/notification.js

```diff
@@ -229,84 +230,72 @@ const Source = GObject.registerClass({
         }
 
         // Fix themed icons
-        if (notificationParams.icon) {
-            let gicon = Gio.Icon.deserialize(notificationParams.icon);
+        if (notification.icon) {
+            let gicon = notification.icon;
 
             if (gicon instanceof Gio.ThemedIcon) {
                 gicon = getIcon(gicon.names[0]);
-                notificationParams.icon = gicon.serialize();
+                notification.icon = gicon.serialize();
             }
         }
 
-        let notification = this._notifications[localId];
+        let cachedNotification = this._notifications[localId];
 
         // Check if this is a repeat
-        if (notification) {
-            notification.requestReplyId = requestReplyId;
+        if (cachedNotification) {
+            cachedNotification.requestReplyId = requestReplyId;
 
             // Bail early If @notificationParams represents an exact repeat
-            const title = notificationParams.title.unpack();
-            const body = notificationParams.body
-                ? notificationParams.body.unpack()
+            const title = notification.title;
+            const body = notification.body
+                ? notification.body
                 : null;
 
-            if (notification.title === title &&
-                notification.bannerBodyText === body) {
+            if (cachedNotification.title === title &&
+                cachedNotification.bannerBodyText === body) {
                 this._notificationPending = false;
                 return;
             }
 
-            notification.title = title;
-            notification.bannerBodyText = body;
+            cachedNotification.title = title;
+            cachedNotification.bannerBodyText = body;
+            cachedNotification.body = body;
 
         // Device Notification
         } else if (idMatch) {
-            notification = this._createNotification(notificationParams);
+            cachedNotification = notification;
 
-            notification.deviceId = deviceId;
-            notification.remoteId = remoteId;
-            notification.requestReplyId = requestReplyId;
+            cachedNotification.deviceId = deviceId;
+            cachedNotification.remoteId = remoteId;
+            cachedNotification.requestReplyId = requestReplyId;
 
-            notification.connect('destroy', (notification, reason) => {
+            cachedNotification.connect('destroy', (notification, reason) => {
                 this._closeGSConnectNotification(notification, reason);
                 delete this._notifications[localId];
             });
 
-            this._notifications[localId] = notification;
+            this._notifications[localId] = cachedNotification;
 
         // Service Notification
         } else {
-            notification = this._createNotification(notificationParams);
-            notification.connect('destroy', (notification, reason) => {
+            cachedNotification.connect('destroy', (notification, reason) => {
```

Comment:
Shouldn't this be `notification`? `cachedNotification` is undefined in this branch

Reply:
I think you're probably right, but TBH that function is way too complex and should probably be like 3 functions.

The move to relying on `cachedNotification` (which may or may not be defined) instead of `notification` (which will always be defined) in so much of the code strikes me as needlessly confusing TBH.

I can see using `cachedNotification` in the first branch (the one taken if it's defined), sure. But since the other branches are _only_ followed when it's **not** defined, why reference it at all? Especially if it's necessary to first set it equal to `notification`. Just use `notification`.

Reply:
Ah yes, sorry I missed that in review. I'm a bit surprised ESLint didn't catch it though; that should certainly be fixed.

Reply:
Never mind, I should've read farther. `cachedNotification` is used because it's needed on lines 287-298.

Reply:
> I think you're probably right, but TBH that function is way too complex and should probably be like 3 functions.

Agreed. For example, already in this patch, `_notificationPending` is no longer correctedly reset in some cases due to the guard statement for existing notification, which from what I can tell is an attempt at reimplementing [`MessageTray.Source.addNotification`](https://gitlab.gnome.org/GNOME/gnome-shell/-/blob/main/js/ui/messageTray.js#L555)